File: android/domain/src/main/kotlin/com/nurseduty/domain/DayKey.kt
package com.nurseduty.domain

import java.time.LocalDate

/**
 * Timezone-independent calendar-day key (yyyymmdd). We use [LocalDate] (no timezone) so a day
 * never shifts or duplicates across timezone/DST changes — the same fix as iOS's DayKey.
 */
object DayKey {
    fun from(date: LocalDate): Int = date.year * 10000 + date.monthValue * 100 + date.dayOfMonth

    fun toLocalDate(key: Int): LocalDate = LocalDate.of(key / 10000, (key / 100) % 100, key % 100)
}

File: Modules/NurseDutyModel/Sources/NurseDutyModel/DayKey.swift
import Foundation

/// Timezone-independent calendar-day key (yyyymmdd). Storing this instead of a wall-clock-midnight
/// `Date` means a day never shifts or duplicates when the device timezone or DST changes.
public enum DayKey {
    public static func from(_ date: Date, _ calendar: Calendar = .current) -> Int {
        let c = calendar.dateComponents([.year, .month, .day], from: date)
        return (c.year ?? 0) * 10000 + (c.month ?? 0) * 100 + (c.day ?? 0)
    }

    /// Reconstruct the start of that calendar day (for display + fire-time math).
    public static func date(_ key: Int, _ calendar: Calendar = .current) -> Date {
        var c = DateComponents()
        c.year = key / 10000
        c.month = (key / 100) % 100
        c.day = key % 100
        return calendar.date(from: c) ?? Date(timeIntervalSinceReferenceDate: 0)
    }
}
